modules/python/python-persist.c

```diff
@@ -127,3 +161,272 @@ python_format_persist_name(const LogPipe *p, const gchar *module, PythonPersistM
+{
+  PyPersist *self =(PyPersist *)s;
+
+  g_free(self->persist_name);
+  self->persist_name = NULL;
+
+  py_slng_generic_dealloc(s);
+}
+
+static PyMemberDef
+py_persist_type_members[] =
+{
+  {"persist_name", T_STRING, offsetof(PyPersist, persist_name), READONLY},
+  {NULL}
+};
+
+static gchar *
+_build_key(PyPersist *self, const gchar *key)
+{
+  return g_strdup_printf("%s" SUBKEY_DELIMITER "%s", self->persist_name, key);
+}
+
+static gchar *
+_lookup_entry(PyPersist *self, const gchar *key, guint8 *type)
+{
+  gchar *query_key = _build_key(self, key);
+
+  gsize size;
+  guint8 version;
+  PersistEntryHandle handle = persist_state_lookup_entry(self->persist_state, query_key, &size, &version);
+  if (!handle)
+    {
+      PyErr_Format(PyExc_KeyError, "Persist has no such key: %s", key);
+      g_free(query_key);
+      return NULL;
+    }
+
+  Entry *entry = persist_state_map_entry(self->persist_state, handle);
+  *type = entry->type;
+  gchar *copy = g_strdup(entry->data);
+  persist_state_unmap_entry(self->persist_state, handle);
+
+  g_free(query_key);
+
+  return copy;
+};
+
+static PersistEntryHandle
+_allocate_persist_entry(PersistState *persist_state, const gchar *key, const gchar *value, gsize value_len)
+{
+  gsize size;
+  guint8 version;
+  PersistEntryHandle handle = persist_state_lookup_entry(persist_state, key, &size, &version);
+  if (handle && size >= value_len)
+    return handle;
+
+  return persist_state_alloc_entry(persist_state, key, value_len);
+}
+
+static gchar *
+_serialize(guint8 type, PyObject *v)
+{
+  if (type == ENTRY_TYPE_STRING)
+    {
+      return g_strdup(_py_get_string_as_string(v));
+    }
+  else if (type == ENTRY_TYPE_LONG)
+    {
+      PyObject *as_str = PyObject_Str(v);
+      gchar *result = g_strdup(_py_get_string_as_string(as_str));
+      Py_DECREF(as_str);
+      return result;
+    }
+  else
+    g_assert_not_reached();
+}
+
+static gboolean
+_store_entry(PyPersist *self, const gchar *key, guint8 type, PyObject *v)
+{
+  gchar *query_key = _build_key(self, key);
+  gchar *value = _serialize(type, v);
+  gsize value_len = strlen(value) + sizeof(type);
+
+  PersistEntryHandle handle = _allocate_persist_entry(self->persist_state, query_key, value, value_len);
+  if (!handle)
+    return FALSE;
+
+  Entry *entry = persist_state_map_entry(self->persist_state, handle);
+  entry->type = type;
+  strcpy(entry->data, value);
+  persist_state_unmap_entry(self->persist_state, handle);
+
+  g_free(value);
+  g_free(query_key);
+
+  return TRUE;
+}
+
+static PyObject *
+_py_persist_type_get(PyObject *o, PyObject *key)
+{
+  PyPersist *self = (PyPersist *)o;
+
+  if (!_py_is_string(key))
+    {
+      PyErr_SetString(PyExc_TypeError, "key is not a string object");
+      return NULL;
+    }
+
+  const gchar *name = _py_get_string_as_string(key);
+  guint8 type;
+  gchar *value = _lookup_entry(self, name, &type);
+
+  if (!value)
+    {
+      PyErr_Format(PyExc_KeyError, "No such name-value pair %s", name);
```

Comment:
Name is leaked

Reply:
Name is stored either in a scratchbuffers, or inside a python object. It should not be freed.

Reply:
ok